### DISCUSSION TITLE
TopologyAppDictionary not populating with all enums used in arrays and parameters, preventing gds from starting up

### DISCUSSION THREAD

**Discussion by apaletta3** [nasa/fprime#1909] [Q&A]
Hey devs,

I'm working on a project using Fprime 3.1.1 and the gds 3.1.4, and I'm not able to start the gds up because it isn't able to parse the dictionary since certain enums that are being used in arrays and parameters that appear in the dictionary don't appear themselves in the dictionary, resulting in a parsing error as such:
`[ERROR] Parsing error: Could not find type SSDL::FormationStatusValue`

This is how the some of these enums appear in the `fpp` files in the project
``` 
enum FormationStatusValue { WORKING, NOT_WORKING, IRRELEVANT, NONE } default IRRELEVANT
array FormationStatuses = [7] FormationStatusValue
array ResponseTableInputs = [20] FormationStatuses
```
Arrays such as these are used in parameters and telem channels for some of the components in the deployment as in the example above where `ResponseTableInputs`  is a parameter. For this example, the dictionary is missing the `FormationStatusValue` enum, but it does include the `FormationStatuses` array and `ResponseTableInputs` array.

Not sure if this helps, but here's an example of how `FormationStatuses` appears in the dict, and since there is no previous definition of `SSDL::FormationStatusValue` as an enum, this is where the error is caused.
```
    <array name="SSDL::FormationStatuses" type="SSDL::FormationStatusValue" type_id="0x34860AEB" size="7" format="%s">
      <defaults>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
        <default value="SSDL::FormationStatusValue::IRRELEVANT"/>
      </defaults>
    </array>
```

Thanks for your help and let me know if there's anything else I should include in this post to help with understanding my issue.

**Comment by LeStarch** [comment]
This looks like it may be a bug.  I'll look into it.

As an aside, Arrays and Serializables cannot be commanded from the GDS at this time.  Thus their use as parameters is somewhat limited.

**Reply by LeStarch** [reply]
I was able to reproduce this behavior.  I'll look into why this is occurring.

**Comment by LeStarch** [comment]
This is a bug in the dictionary generation tool.  Namely it does not support the following types:

1. Serializable with Serializable members: silent error
2. Array with Arrary, Enum, or Serializable member

As a work around you could define unused telemetry channels with the nested types.  This makes them first class dictionary memebers, they are generated, and then can be used in nested types.

e.g.

```
telemetry my_real_channel: ResponseTableInputs 0
telemetry __unused_1__: SSDL.FormationStatusValue id 97
telemetry __unused_2__: SSDL.FormationStatuses id 98
```
This will force the types to be added to the dictionary. These types still cannot be commanded, but this allows them as telemetry items.

**Comment by ebenezerarunkumar** [comment]
I was having a similar error where I have the following types defined in my component FPP:

``` 
enum SubsystemState { ON, OFF } default OFF
struct AllSubsystemStates { propState: SubsystemState, paibState: SubsystemState, xlinkState: SubsystemState, scienceState: SubsystemState }
struct SubsystemStatesPerMode { MANUAL: AllSubsystemStates, STANDBY: AllSubsystemStates, TRANSFER_IN: AllSubsystemStates, TRANSFER_OUT: AllSubsystemStates, SCIENCE: AllSubsystemStates, SAFE: AllSubsystemStates, ESCAPE: AllSubsystemStates, NONE: AllSubsystemStates }
  ```
  When I try to integrate this component into my deployment, I get the following error during build: 
  
  ```
  Parsing Enum SubsystemState
Parsing Serializable SubsystemStatesPerMode
/Users/earunkumar/fprime-devel/visors-software/Hsa/build-fprime-automatic-native/Hsa/PayloadStateMachine/SubsystemStatesPerModeSerializableAi.xml: Can only include one level of serializable for dictionaries
Traceback (most recent call last):
  File "/Users/earunkumar/fprime-devel/visors-software/fprime/Autocoders/Python/bin/codegen.py", line 1334, in <module>
    main()
  File "/Users/earunkumar/fprime-devel/visors-software/fprime/Autocoders/Python/bin/codegen.py", line 1283, in main
    generate_topology(
  File "/Users/earunkumar/fprime-devel/visors-software/fprime/Autocoders/Python/bin/codegen.py", line 421, in generate_topology
    top_dict_gen.check_for_serial_xml()
  File "/Users/earunkumar/fprime-devel/visors-software/fprime/Autocoders/Python/src/fprime_ac/utils/TopDictGenerator.py", line 42, in check_for_serial_xml
    raise Exception(
Exception: /Users/earunkumar/fprime-devel/visors-software/Hsa/build-fprime-automatic-native/Hsa/PayloadStateMachine/SubsystemStatesPerModeSerializableAi.xml: Can only include one level of serializable for dictionaries
make[3]: *** [Hsa/Top/HsaTopologyAppDictionary.xml] Error 255
make[2]: *** [Hsa/Top/CMakeFiles/Hsa_Top.dir/all] Error 2
make[1]: *** [CMakeFiles/Hsa.dir/rule] Error 2
make: *** [Hsa] Error 2
[ERROR] CMake erred with return code 2
```


It looks like fprime does not like the nested serializable

**Reply by LeStarch** [reply]
Nested structures within dictionary items (Commands, Events, Telemetry) are (currently) not supported.

Ironically, nesting the structure within an array of size one does work with the above work around.

**Reply by LeStarch** [reply]
We are looking into ironing all this out.
